### PR TITLE
Fix cancellation race condition

### PR DESCRIPTION
**Pull request by 8W9aG**
* Currently when the user cancels the cancellation token we receive this delegate callback in the SPTDataLoaderService
* The SPTDataLoaderService then proceeds to cancel the associated data task
* This would then asynchronously cancel the request, once cancelled that message was forwarded up to the SPTDataLoader which would then alert the delegate
* The problem occurs because of the asynchronousity of the data task cancellation, we could theoretically receive an error or success message before we cancel the data task despite the fact that the cancellation came first
* We will solve this problem by cancelling requests in SPTDataLoader instead and disallowing them  to bubbled up to the UI immediately once cancelled, and perform the actual cancel further down the stack

### CODE REVIEW

**Comment by 8W9aG** [comment]
@nataliq

**Comment by 8W9aG** [comment]
Tests are coming, verify whether this architecture sounds sane first

File: SPTDataLoader/SPTDataLoaderService.m [hunk 1]
```diff
@@ -187,24 +184,34 @@ - (void)cancelAllLoads
...
+}
+
+- (void)requestResponseHandler:(id<SPTDataLoaderRequestResponseHandler>)requestResponseHandler
+                 cancelRequest:(SPTDataLoaderRequest *)request
+{
+    NSArray *handlers = nil;
+    @synchronized(self.handlers) {
+        handlers = [self.handlers copy];
+    }
+    for (SPTDataLoaderRequestTaskHandler *handler in handlers) {
```
Comment (nataliq) [inline SPTDataLoader/SPTDataLoaderService.m]: Shouldn't you use the provided `requestResponseHandler` instead?

File: SPTDataLoader/SPTDataLoaderService.m [hunk 2]
```diff
@@ -187,24 +184,34 @@ - (void)cancelAllLoads
 
 #pragma mark SPTDataLoaderRequestResponseHandlerDelegate
 
-- (nullable id<SPTDataLoaderCancellationToken>)requestResponseHandler:(id<SPTDataLoaderRequestResponseHandler>)requestResponseHandler
-                                                       performRequest:(SPTDataLoaderRequest *)request
+- (void)requestResponseHandler:(id<SPTDataLoaderRequestResponseHandler>)requestResponseHandler
```
Comment (nataliq) [inline SPTDataLoader/SPTDataLoaderService.m]: It makes more sense to combine `requestResponseHandler:performRequest:` and `performRequest:requestResponseHandler:`, this way it is confusing.

File: SPTDataLoader/SPTDataLoaderService.m (hunk 1 above)
Comment (8W9aG) [inline SPTDataLoader/SPTDataLoaderService.m]: A request task handler isn't the same as a requestResponseHandler

File: SPTDataLoader/SPTDataLoaderService.m (hunk 2 above)
Comment (8W9aG) [inline SPTDataLoader/SPTDataLoaderService.m]: The separation is needed due to how the authentication logic works

**Comment by KeenanCassidy** [comment]
👍

**Pull request merged**